I.1 Univariate and Multivariate EDA

In [1]:
#First, we mount Google Drive to Colab.
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [8]:
#This code exists only to check the path of the CSV file and should be commented out after execution
#!ls
#%cd drive/MyDrive/
#!ls

In [33]:
#Now, we import the necessary libraries for EDA and classification
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split


In [34]:
#Read in the CSV using Pandas.
df = pd.read_csv('online_shoppers_intention.csv.gz')

In [35]:
#See what data looks like
df.head()


Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

In [36]:
df.tail()

Administrative  Administrative_Duration  Informational  \
12325               3                    145.0              0   
12326               0                      0.0              0   
12327               0                      0.0              0   
12328               4                     75.0              0   
12329               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
12325                     0.0              53              1783.791667   
12326                     0.0               5               465.750000   
12327                     0.0               6               184.250000   
12328                     0.0              15               346.000000   
12329                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
12325     0.007143   0.029031   12.241717         0.0   Dec                 4   
12326     0.000000   0.021333    0.000000         0.0   Nov                 3   
12327     0.083333   0.086667    0.000000         0.0   Nov                 3   
12328     0.000000   0.021053    0.000000         0.0   Nov                 2   
12329     0.000000   0.066667    0.000000         0.0   Nov                 3   

       Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
12325        6       1            1  Returning_Visitor     True    False  
12326        2       1            8  Returning_Visitor     True    False  
12327        2       1           13  Returning_Visitor     True    False  
12328        2       3           11  Returning_Visitor    False    False  
12329        2       1            2        New_Visitor     True    False

In [37]:
#print(df[df['Informational']>0])  #This was to check if "Informational" was a binary variable, it is not
#print(df[df['Informational_Duration']>0.0]) #This was to check the kind of values Informational_Duration can have.  
#df.shape
#There are 12,330 data points, with 18 features.  The .head() and .tail() functions show us that 

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12330 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12330 non-null  int64  
 1   Administrative_Duration  12330 non-null  float64
 2   Informational            12330 non-null  int64  
 3   Informational_Duration   12330 non-null  float64
 4   ProductRelated           12330 non-null  int64  
 5   ProductRelated_Duration  12330 non-null  float64
 6   BounceRates              12330 non-null  float64
 7   ExitRates                12330 non-null  float64
 8   PageValues               12330 non-null  float64
 9   SpecialDay               12330 non-null  float64
 10  Month                    12330 non-null  object 
 11  OperatingSystems         12330 non-null  int64  
 12  Browser                  12330 non-null  int64  
 13  Region                   12330 non-null  int64  
 14  TrafficType           

In [92]:
df_noncategorical = df.drop(['OperatingSystems', 'Browser', 'Region', 'TrafficType', 'VisitorType','Weekend', 'Revenue', 'SpecialDay', 'Month', 'PageValues'], axis=1)
df_noncategorical.describe()

 #This method does what we need for univariate EDA.  The statistics are not helpful for categorical variables such as "OpreatingSystems," "Browser", "Region", "TrafficType", "VisitorType", "Weekend", and "Revenue"

Administrative  Administrative_Duration  Informational  \
count    12330.000000             12330.000000   12330.000000   
mean         2.315166                80.818611       0.503569   
std          3.321784               176.779107       1.270156   
min          0.000000                 0.000000       0.000000   
25%          0.000000                 0.000000       0.000000   
50%          1.000000                 7.500000       0.000000   
75%          4.000000                93.256250       0.000000   
max         27.000000              3398.750000      24.000000   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
count            12330.000000    12330.000000             12330.000000   
mean                34.472398       31.731468              1194.746220   
std                140.749294       44.475503              1913.669288   
min                  0.000000        0.000000                 0.000000   
25%                  0.000000        7.000000               184.137500   
50%                  0.000000       18.000000               598.936905   
75%                  0.000000       38.000000              1464.157214   
max               2549.375000      705.000000             63973.522230   

        BounceRates     ExitRates  
count  12330.000000  12330.000000  
mean       0.022191      0.043073  
std        0.048488      0.048597  
min        0.000000      0.000000  
25%        0.000000      0.014286  
50%        0.003112      0.025156  
75%        0.016813      0.050000  
max        0.200000      0.200000

In [91]:
#Complete univariate EDA by getting appropriate statistics for the remaining features
df_categorical = df.drop(['Administrative', 'Administrative_Duration', 'Informational', 'Informational_Duration', 'ProductRelated', 'ProductRelated_Duration', 'BounceRates', 'ExitRates',  'Revenue'], axis = 1)
df_categorical.astype('object').describe()

PageValues  SpecialDay  Month  OperatingSystems  Browser  Region  \
count      12330.0     12330.0  12330             12330    12330   12330   
unique      2704.0         6.0     10                 8       13       9   
top            0.0         0.0    May                 2        2       1   
freq        9600.0     11079.0   3364              6601     7961    4780   

        TrafficType        VisitorType Weekend  
count         12330              12330   12330  
unique           20                  3       2  
top               2  Returning_Visitor   False  
freq           3913              10551    9462

In [70]:
print("We summarize the relevant non-categorical features.")
print("Individuals visit an average of 2.3 administrative pages, a median of 1.0, and stay for an average of 80.8 minutes.")
print("Individuals visit an average of 0.50 informational pages, a median of 0.0, and stay for an average of 34.5 minutes.")
print("Individuals visit an average of 31.7 produce-related pages, a median of 18.0 minutes, and stay for an average of 1194.7 minutes.")
print("The medians are consistently lower than the means, so the distribution is not normal.")
print("")
print("We summarize the relevant categorical features.")
print("Operating System 2 is the most common operating system, acting as the operating system for " + str(100*6601/12330) + "% of all visitors.  There are 8 unique operatin systems")
print("Browser 2 is the most common browser for all visitors, acting as the browser for " + str(100*7961/12330) + "% of all visitors.  There are 13 unique browsers")
print("Region 1 is the most common region for all visitors, acting as the region for " + str(100*4780/12330) + "% of all visitors.  There are 9 unique regions")
print("Returning visitors are " + str(100*10551/12330) + "% of all visitors.  There are 3 unique visitor types.")
print("TrafficType 2 accounts for " + str(100*3913/12330) + "% of all visits.  There are 9 unique traffic types.")
print("The weekday accounts for " + str(100*9462/12330) + "% of all visits.  There is either the weekend or the weekday.")

We summarize the relevant non-categorical features.
Individuals visit an average of 2.3 administrative pages, a median of 1.0, and stay for an average of 80.8 minutes.
Individuals visit an average of 0.50 informational pages, a median of 0.0, and stay for an average of 34.5 minutes.
Individuals visit an average of 31.7 produce-related pages, a median of 18.0 minutes, and stay for an average of 1194.7 minutes.
The medians are consistently lower than the means, so the distribution is not normal.

We summarize the relevant categorical features.
Operating System 2 is the most common operating system, acting as the operating system for 53.53609083536091% of all visitors.  There are 8 unique operatin systems
Browser 2 is the most common browser for all visitors, acting as the browser for 64.56609894566098% of all visitors.  There are 13 unique browsers
Region 1 is the most common region for all visitors, acting as the region for 38.767234387672346% of all visitors.  There are 9 unique region

In [96]:
#It would be useful to get statistics on correlations.

df.corr()

Administrative  Administrative_Duration  \
Administrative                 1.000000                 0.601583   
Administrative_Duration        0.601583                 1.000000   
Informational                  0.376850                 0.302710   
Informational_Duration         0.255848                 0.238031   
ProductRelated                 0.431119                 0.289087   
ProductRelated_Duration        0.373939                 0.355422   
BounceRates                   -0.223563                -0.144170   
ExitRates                     -0.316483                -0.205798   
PageValues                     0.098990                 0.067608   
SpecialDay                    -0.094778                -0.073304   
OperatingSystems              -0.006347                -0.007343   
Browser                       -0.025035                -0.015392   
Region                        -0.005487                -0.005561   
TrafficType                   -0.033561                -0.014376   
Weekend                        0.026417                 0.014990   
Revenue                        0.138917                 0.093587   

                         Informational  Informational_Duration  \
Administrative                0.376850                0.255848   
Administrative_Duration       0.302710                0.238031   
Informational                 1.000000                0.618955   
Informational_Duration        0.618955                1.000000   
ProductRelated                0.374164                0.280046   
ProductRelated_Duration       0.387505                0.347364   
BounceRates                  -0.116114               -0.074067   
ExitRates                    -0.163666               -0.105276   
PageValues                    0.048632                0.030861   
SpecialDay                   -0.048219               -0.030577   
OperatingSystems             -0.009527               -0.009579   
Browser                      -0.038235               -0.019285   
Region                       -0.029169               -0.027144   
TrafficType                  -0.034491               -0.024675   
Weekend                       0.035785                0.024078   
Revenue                       0.095200                0.070345   

                         ProductRelated  ProductRelated_Duration  BounceRates  \
Administrative                 0.431119                 0.373939    -0.223563   
Administrative_Duration        0.289087                 0.355422    -0.144170   
Informational                  0.374164                 0.387505    -0.116114   
Informational_Duration         0.280046                 0.347364    -0.074067   
ProductRelated                 1.000000                 0.860927    -0.204578   
ProductRelated_Duration        0.860927                 1.000000    -0.184541   
BounceRates                   -0.204578                -0.184541     1.000000   
ExitRates                     -0.292526                -0.251984     0.913004   
PageValues                     0.056282                 0.052823    -0.119386   
SpecialDay                    -0.023958                -0.036380     0.072702   
OperatingSystems               0.004290                 0.002976     0.023823   
Browser                       -0.013146                -0.007380    -0.015772   
Region                        -0.038122                -0.033091    -0.006485   
TrafficType                   -0.043064                -0.036377     0.078286   
Weekend                        0.016092                 0.007311    -0.046514   
Revenue                        0.158538                 0.152373    -0.150673   

                         ExitRates  PageValues  SpecialDay  OperatingSystems  \
Administrative           -0.316483    0.098990   -0.094778         -0.006347   
Administrative_Duration  -0.205798    0.067608   -0.073304         -0.007343   
Informational            -0.163666    0.048632   -0.048219         -0.009527   
Informational_Duration   -0.105276    0.030861 

In [ ]:
#I see no significant correlations between features, with the exvption of correlations between type of page visited and duration of time spent on such pages.  Revenue is weakly correlated with most features, showing correlations of around 0.15 with number of administrative pages visited and number of product-related pages visited.
#There is a correlation in the opposite direction between revenue and bounce rate (-0.15) and revenue and exit rate(-0.21).The exception is that revenue has nearly a 0.50 correlation with page values.

I.2 Data Pre-Processing and Cross-Validation

In [97]:
#Now, we manually remove features from the dataframe that inspection would suggest are nonessential.  Based on correlations, it would make sense to trian on 'Administrative', 'Product_Related', 'Page Value', 'Bounce Rates', and 'Exit Rates,'  This is five features.
df_training = df.drop(['Month', 'VisitorType','Administrative_Duration', 'Informational', 'Informational_Duration', 'ProductRelated', 'ProductRelated_Duration', 'SpecialDay', 'OperatingSystems', 'Browser', 'Region', 'TrafficType', 'Weekend'], axis=1)
df_training.head()

Administrative  BounceRates  ExitRates  PageValues  Revenue
0               0         0.20       0.20         0.0    False
1               0         0.00       0.10         0.0    False
2               0         0.20       0.20         0.0    False
3               0         0.05       0.14         0.0    False
4               0         0.02       0.05         0.0    False

In [ ]:
#Now, we correct for the class imbalance w.r.t. Revenue using SMOTE

In [108]:
X = df_training.drop('Revenue', axis=1)
X.head()
y = df.Revenue
y.head()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(8261, 4)
(8261,)
(4069, 4)
(4069,)


I.3 Modeling

In [125]:
#Here, we perform and evaluate logistic regression

from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.metrics import confusion_matrix

clf = LogisticRegression(random_state=0).fit(X_train, y_train)
y_pred = clf.predict(X_test)
score = metrics.accuracy_score(y_test,y_pred)

confusion_matrix(y_test, y_pred)
print("Accuracy is " + str(100*score) +"%.")
print(confusion_matrix(y_test, y_pred))

Accuracy is 88.15433767510444%.
[[3368   68]
 [ 414  219]]


In [130]:
#Here, we use kNN and compare with the logistic regressio.
#Selection of k is important here.  We will loop over different ks and select based on highest accuracy.

from sklearn.neighbors import KNeighborsClassifier

scores = []
for k in range(1,50):
  knn = KNeighborsClassifier(n_neighbors = k)
  knn.fit(X_train,y_train)
  y_pred = knn.predict(X_test)
  score = metrics.accuracy_score(y_test, y_pred)
  scores.append(score)
max_score = max(scores)
max_k = scores.index(max_score)
print("Max k is " + str(max_k))
knn = KNeighborsClassifier(n_neighbors = max_k)
knn.fit(X_train,y_train)
y_pred = knn.predict(X_test)
score = metrics.accuracy_score(y_test, y_pred)
scores.append(score)
print("Accuracy is " + str(100*score) +"%.")
print(confusion_matrix(y_test, y_pred))

Max k is 36
Accuracy is 89.1373801916933%.
[[3282  154]
 [ 288  345]]


II. Customer Segmentation

In [142]:
#Here, we perform k-means clustering with scikit-learn
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=2, random_state=0).fit(X)
print(len(kmeans.cluster_centers_))
np.bincount(kmeans.labels_)

new_column = kmeans.predict(X)
data = df.copy()
data['Cluster'] = new_column
data.head()
data_cluster0 = data[data['Cluster']==0]
data_cluster1 = data[data['Cluster']!=0]
data_cluster0.describe()


2


Administrative  Administrative_Duration  Informational  \
count    11639.000000             11639.000000   11639.000000   
mean         2.279835                79.345886       0.498067   
std          3.311849               176.581072       1.263243   
min          0.000000                 0.000000       0.000000   
25%          0.000000                 0.000000       0.000000   
50%          1.000000                 5.000000       0.000000   
75%          4.000000                90.937500       0.000000   
max         27.000000              3398.750000      24.000000   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
count            11639.000000    11639.000000             11639.000000   
mean                34.450951       31.603832              1187.542369   
std                142.373251       45.006727              1941.881734   
min                  0.000000        0.000000                 0.000000   
25%                  0.000000        7.000000               170.375000   
50%                  0.000000       17.000000               577.000000   
75%                  0.000000       38.000000              1448.840476   
max               2549.375000      705.000000             63973.522230   

        BounceRates     ExitRates    PageValues    SpecialDay  \
count  11639.000000  11639.000000  11639.000000  11639.000000   
mean       0.023328      0.044747      2.290921      0.064284   
std        0.049642      0.049430      6.266873      0.203227   
min        0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.014944      0.000000      0.000000   
50%        0.003750      0.026667      0.000000      0.000000   
75%        0.018182      0.050000      0.000000      0.000000   
max        0.200000      0.200000     34.348571      1.000000   

       OperatingSystems       Browser        Region   TrafficType  Cluster  
count      11639.000000  11639.000000  11639.000000  11639.000000  11639.0  
mean           2.122863      2.345133      3.149154      4.060143      0.0  
std            0.898257      1.684023      2.398929      4.006417      0.0  
min            1.000000      1.000000      1.000000      1.000000      0.0  
25%            2.000000      2.000000      1.000000      2.000000      0.0  
50%            2.000000      2.000000      3.000000      2.000000      0.0  
75%            3.000000      2.000000      4.000000      4.000000      0.0  
max            8.000000     13.000000      9.000000     20.000000      0.0

In [141]:
data_cluster1.describe()

Administrative  Administrative_Duration  Informational  \
count      691.000000               691.000000     691.000000   
mean         2.910275               105.624755       0.596237   
std          3.433095               178.391742       1.379115   
min          0.000000                 0.000000       0.000000   
25%          0.000000                 0.000000       0.000000   
50%          2.000000                48.187500       0.000000   
75%          4.000000               131.750000       0.000000   
max         22.000000              1668.500000      14.000000   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
count              691.000000      691.000000               691.000000   
mean                34.833644       33.881331              1316.085762   
std                109.927253       34.267707              1348.018525   
min                  0.000000        1.000000                15.600000   
25%                  0.000000       13.000000               502.125000   
50%                  0.000000       24.000000               922.400606   
75%                  0.000000       41.000000              1614.687500   
max               1021.600000      318.000000             13717.402050   

       BounceRates   ExitRates  PageValues  SpecialDay  OperatingSystems  \
count   691.000000  691.000000  691.000000  691.000000        691.000000   
mean      0.003041    0.014876   66.498590    0.013314          2.143271   
std       0.007514    0.012009   40.013133    0.089098          1.109130   
min       0.000000    0.000000   34.436019    0.000000          1.000000   
25%       0.000000    0.006733   43.111473    0.000000          2.000000   
50%       0.000000    0.012361   54.248864    0.000000          2.000000   
75%       0.002660    0.020000   75.172472    0.000000          2.000000   
max       0.060000    0.075000  361.763742    1.000000          8.000000   

          Browser      Region  TrafficType  Cluster  
count  691.000000  691.000000   691.000000    691.0  
mean     2.558611    3.117221     4.228654      1.0  
std      2.194980    2.447568     4.328850      0.0  
min      1.000000    1.000000     1.000000      1.0  
25%      2.000000    1.000000     2.000000      1.0  
50%      2.000000    3.000000     2.000000      1.0  
75%      2.000000    4.000000     4.000000      1.0  
max     13.000000    9.000000    20.000000      1.0

In [143]:
data_cluster0.astype('object').describe()

Administrative  Administrative_Duration  Informational  \
count            11639                  11639.0          11639   
unique              27                   3144.0             17   
top                  0                      0.0              0   
freq              5557                   5684.0           9179   

        Informational_Duration  ProductRelated  ProductRelated_Duration  \
count                  11639.0           11639                  11639.0   
unique                  1188.0             308                   8961.0   
top                        0.0               1                      0.0   
freq                    9393.0             620                    755.0   

        BounceRates  ExitRates  PageValues  SpecialDay  Month  \
count       11639.0    11639.0     11639.0     11639.0  11639   
unique       1834.0     4605.0      2029.0         6.0     10   
top             0.0        0.2         0.0         0.0    May   
freq         5028.0      710.0      9600.0     10406.0   3206   

        OperatingSystems  Browser  Region  TrafficType        VisitorType  \
count              11639    11639   11639        11639              11639   
unique                 8       13       9           20                  3   
top                    2        2       1            2  Returning_Visitor   
freq                6171     7533    4499         3591              10075   

       Weekend Revenue  Cluster  
count    11639   11639    11639  
unique       2       2        1  
top      False   False        0  
freq      8940   10270    11639

In [144]:
data_cluster1.astype('object').describe()

Administrative  Administrative_Duration  Informational  \
count              691                    691.0            691   
unique              19                    407.0             10   
top                  0                      0.0              0   
freq               211                    219.0            520   

        Informational_Duration  ProductRelated  ProductRelated_Duration  \
count                    691.0             691                    691.0   
unique                   145.0             111                    687.0   
top                        0.0              13                    268.0   
freq                     532.0              29                      2.0   

        BounceRates  ExitRates  PageValues  SpecialDay Month  \
count         691.0      691.0     691.000       691.0   691   
unique        143.0      372.0     675.000         6.0    10   
top             0.0        0.0      53.988         0.0   Nov   
freq          490.0       29.0       6.000       673.0   211   

        OperatingSystems  Browser  Region  TrafficType        VisitorType  \
count                691      691     691          691                691   
unique                 6       10       9           15                  3   
top                    2        2       1            2  Returning_Visitor   
freq                 430      428     281          322                476   

       Weekend Revenue  Cluster  
count      691     691      691  
unique       2       2        1  
top      False    True        1  
freq       522     539      691

Ultimately, we can say that the two clusters generated by k-means represent those customers who purchased something and those who did not.  539 out of the 691 members of cluster 1 (78%) purchased something.  1369/11639 (11.8%) purchased something.  A major difference between the clusters seems to be lower bounce rates in cluster 1 and more administrative pages visited.

III. Semi-Supervised Learning

I began this task and attempted to perform label propagation.  I believe that I was able to do so successfully for 'Revenue', but due to difficulties and time constraints, could not complete the task.  I am uploading the notebook as is.

In [146]:
data.info()
data.astype('object').describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12330 entries, 0 to 12329
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12330 non-null  int64  
 1   Administrative_Duration  12330 non-null  float64
 2   Informational            12330 non-null  int64  
 3   Informational_Duration   12330 non-null  float64
 4   ProductRelated           12330 non-null  int64  
 5   ProductRelated_Duration  12330 non-null  float64
 6   BounceRates              12330 non-null  float64
 7   ExitRates                12330 non-null  float64
 8   PageValues               12330 non-null  float64
 9   SpecialDay               12330 non-null  float64
 10  Month                    12330 non-null  object 
 11  OperatingSystems         12330 non-null  int64  
 12  Browser                  12330 non-null  int64  
 13  Region                   12330 non-null  int64  
 14  TrafficType           

Administrative  Administrative_Duration  Informational  \
count            12330                  12330.0          12330   
unique              27                   3335.0             17   
top                  0                      0.0              0   
freq              5768                   5903.0           9699   

        Informational_Duration  ProductRelated  ProductRelated_Duration  \
count                  12330.0           12330                  12330.0   
unique                  1258.0             311                   9551.0   
top                        0.0               1                      0.0   
freq                    9925.0             622                    755.0   

        BounceRates  ExitRates  PageValues  SpecialDay  Month  \
count       12330.0    12330.0     12330.0     12330.0  12330   
unique       1872.0     4777.0      2704.0         6.0     10   
top             0.0        0.2         0.0         0.0    May   
freq         5518.0      710.0      9600.0     11079.0   3364   

        OperatingSystems  Browser  Region  TrafficType        VisitorType  \
count              12330    12330   12330        12330              12330   
unique                 8       13       9           20                  3   
top                    2        2       1            2  Returning_Visitor   
freq                6601     7961    4780         3913              10551   

       Weekend Revenue  Cluster  
count    12330   12330    12330  
unique       2       2        2  
top      False   False        0  
freq      9462   10422    11639

In [233]:
#This cell is intended to merely verify the October, November, and December records are indeed filled with NaN values, as the specification indicates.  We check on October.
#Then, we create different datasets to align with the specifications.
data['Month'].unique()
segmented_data10 = data[data['Month']== ('Oct')]
segmented_data10.describe()
print(data['Month'].unique())
segmented_data1 = data[data['Month']== ('Jan')] #Nothing here
segmented_data2 = data[data['Month']== ('Feb')]
segmented_data3 = data[data['Month']== ('Mar')]
segmented_data4 = data[data['Month']== ('Apr')] #Nothing here
segmented_data5 = data[data['Month']== ('May')]
segmented_data6 = data[data['Month']== ('June')]
segmented_data7 = data[data['Month']== ('Jul')]
segmented_data8 = data[data['Month']== ('Aug')]
segmented_data9 = data[data['Month']== ('Sep')]
segmented_data11 = data[data['Month']== ('Nov')]
segmented_data12 = data[data['Month']== ('Dec')]

test = pd.concat([segmented_data2, segmented_data3])
unlabeled = pd.concat([segmented_data10, segmented_data11, segmented_data12])
labeled = pd.concat([segmented_data5, segmented_data6, segmented_data7, segmented_data8, segmented_data9])

print(labeled['Month'].unique())
print(unlabeled['Month'].unique())
print(test['Month'].unique())






['Feb' 'Mar' 'May' 'Oct' 'June' 'Jul' 'Aug' 'Nov' 'Sep' 'Dec']
['May' 'June' 'Jul' 'Aug' 'Sep']
['Oct' 'Nov' 'Dec']
['Feb' 'Mar']


In [219]:
#We do label propagation
from sklearn.semi_supervised import LabelPropagation
label_prop_model = LabelPropagation()
labels = np.copy(labeled['Revenue'])

rng = np.random.RandomState(42)
random_unlabeled_points = rng.rand(len(labeled['Revenue'])) < 0.15  #Attempt to account for imbalance of dataset
labels[random_unlabeled_points] = 1


labeled_data = labeled.drop(['Revenue'], axis = 1)
labeled_data = labeled_data.drop(['Month', 'VisitorType'], axis=1)
unlabeled = unlabeled.drop(['Month', 'VisitorType', 'Revenue'], axis=1)
labeled_data.head()
labeled_data = np.asarray(labeled_data)
label_prop_model.fit(labeled_data, labels)
unlabeled_revenue_pred = label_prop_model.predict(unlabeled)

print(unlabeled_revenue_pred)


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:444: UserWarning: X has feature names, but LabelPropagation was fitted without feature names
  f"X has feature names, but {self.__class__.__name__} was fitted without"


[False False False ... False False False]


/usr/local/lib/python3.7/dist-packages/sklearn/semi_supervised/_label_propagation.py:222: RuntimeWarning: invalid value encountered in true_divide
  probabilities /= normalizer


In [220]:
unlabeled = pd.DataFrame(unlabeled)
labeled_data = pd.DataFrame(labeled)
labels = pd.DataFrame(labels)

unlabeled['Revenue'] = unlabeled_revenue_pred
labeled_data['Revenue'] = labels

semisupervised_data = pd.concat([labeled_data, unlabeled])


In [229]:
labeled_data.tail()

Administrative  Administrative_Duration  Informational  \
7967               3                54.000000              4   
7969               0                 0.000000              1   
7975               6               142.911111              0   
7976              12               975.646667              4   
7977               4                89.600000              2   

      Informational_Duration  ProductRelated  ProductRelated_Duration  \
7967                   32.00               4                46.000000   
7969                   13.00              27              1620.024762   
7975                    0.00              13               264.339682   
7976                  276.08              53              2822.623603   
7977                   12.00              13               182.500000   

      BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
7967     0.000000   0.022222         0.0         0.0   Sep                 1   
7969     0.050000   0.085714         0.0         0.0   Sep                 5   
7975     0.000000   0.034510         0.0         0.0   Sep                 1   
7976     0.006667   0.017015         0.0         0.0   Sep                 3   
7977     0.000000   0.014815         0.0         0.0   Sep                 2   

      Browser  Region  TrafficType        VisitorType  Weekend Revenue  \
7967        1       1            5        New_Visitor    False     NaN   
7969       11       4            1  Returning_Visitor    False     NaN   
7975        1       3            5  Returning_Visitor    False     NaN   
7976        2       8            2  Returning_Visitor     True     NaN   
7977        2       1            2  Returning_Visitor    False     NaN   

      Cluster  
7967        0  
7969        0  
7975        0  
7976        0  
7977        0

ValueError: ignored